### ISSUE TITLE
Types

### ISSUE THREAD

**Issue by patrickhilker** [InterNetX/js-domainrobot-sdk#6]
Just discovered this lib. Would it be possible to provide types to use those in typescript?

**Comment by adelholtz** [comment]
Hi,

we have added this to our implementation list and will give mention it in this issue ass soon as it is implemented.

Have a nice day.

**Comment by patrickhilker** [comment]
Do you have any plans as to when the types could be delivered?
I'm just curious if it's worth the wait.

**Comment by adelholtz** [comment]
Hi
it may be released within the next two weeks. I can't give a more specific ETA right now because of rather high workload right now but we haven't forgotten about it and strive to release the types as soon as possible :)
Hope this answer is helpful for you.

**Comment by patrickhilker** [comment]
That is awesome, thank you in advance! 👍

**Comment by adelholtz** [comment]
Hi

i have released an incomplete preview version of the definition file. Please take a look and tell me if this is enough for your needs.
Have a nice day.

**Comment by patrickhilker** [comment]
The preview is perhaps a first small step in the right direction.

However, the preview currently contains very little additional information about the types. For example, an 'object' type is almost useless because it doesn't specify expected keys and their types, and that's what it's all about. The return types are nice, but the classes are all empty, so they have no added value. Among other things, all methods are missing in the classes.

Of course it's just a preview - it's okay for that. If you add the additional information, it will be great. Thank you for the effort, I'm looking forward to further improvements!

**Comment by adelholtz** [comment]
Hi

i just release a new more complete version of the definition file. Hopefully this is more what you expected. 
Have a nice day.

**Comment by patrickhilker** [comment]
Hello,

the current version is much better! To improve would be here:

- Please do not use `object` or `any` as type at any point - it is completely useless. Especially with the result it would be nice to have types. This makes programming a lot easier, because the code documents itself and you don't have to check all the time what kind of result you can expect.

- Almost all interface keys are optional. Is that really how it is meant? 

- in some places `[String]` or similar is used as type. This would mean "an array with exactly one string item". I suppose you wanted to use `string[]`?

These headwords caught my eye while looking at it briefly. I look forward to further updates! :-)

**Comment by adelholtz** [comment]
HI,

thank you for the feedback!

> Almost all interface keys are optional. Is that really how it is meant?

Sadly right now, this is meant to be this way. Since there are many tasks that use the same Models certain tasks have certain mandatory fields which others do not have. This is especially true for certificate and domain tasks. Because of this many fields are mandatory in some but optional in other cases which makes it hard to accurately describe a Model/Class without using optional keys in almost all cases.
There are certain keys which are mandatory in all use cases but i will have to cross check which those are, so for now this is the best i can do.

> in some places [String] or similar is used as type. This would mean "an array with exactly one string item". I suppose you wanted to use string[]?
Thank you for the information i didn't know that i will try to correct it as fast as possible.

I am quite new to typescript so any feedback and help is greatly appreciated!

Have a nice day.

**Comment by adelholtz** [comment]
Updated the definitions. -> 1.2.1

**Comment by patrickhilker** [comment]
Can you please find some way to type the responses? As a developer, it's very difficult to work with them, because you have to check the documentation all the time. With types, the code would document itself wonderfully.

The types must be available in some form - after all, there is also the swagger definition of your API which contains the response types. We have used this to generate a client from it, but now you get over 600 typescript errors when you try to transpile the fresh generated client.

https://github.com/InterNetX/internetx-swagger-files

**Comment by adelholtz** [comment]
Hi

do you mean this?

```typescript
export declare class ContactService {
   create(): DomainRobotResult;
    update(): DomainRobotResult;
    list(): DomainRobotResult;
    info(): DomainRobotResult;
    delete(): DomainRobotResult;
} 
```

You need a better definition for "DomainRobotResult" ?

Sorry that you didn't get an answer on the swagger-files issue I will talk with my co-worker about this. :)

I do believe that the swagger file is broken in a way as well, because I had problems creating my php/js/ruby swagger client from the definition as well. I will look into it when I have time and try to get back to you.

Have a nice day.

**Comment by patrickhilker** [comment]
Exactly! Thanks in advance!

**Comment by adelholtz** [comment]
Hi

since I have little experience in working with typescript I wanted to ask if this definition would work for you:

```
export declare class ContactService {
	create(): DomainRobotResult<JsonResponseDataContact, Number>;
}

export declare class DomainRobotResult<Result, Number> {
	constructor(result: Result, status: Number);
	isValid(): Boolean;
	isValidResponse(): Boolean;
}

export interface DomainRobotResult<Result, Number> {
	result: Result;
	status: Number;
}

export interface ResponseStatus {
	code: String;
	text: String;
	type: String;
}

export interface ResponseObject {
	type: String;
	value: String;
	summary?: Number;
	data?: String;
}

export interface Result {
	stid: String;
	status: ResponseStatus;
	object?: ResponseObject;
	ctid?: String;
	data: Array<object>;
}

export interface JsonResponseDataContact extends Result {
	data: DomainRobotModels.Contact[];
}

```

The tests I did with this suggest that it should work but since I don't actually work with typescript I am not sure if this is what you would expect. So I would be grateful if you could comment on this. Thank you :)

**Comment by adelholtz** [comment]
Version 1.3.1 should fix all type related problems and errors.
I implemented a very basic example in typescript to test the definitions file and it seems to work fine.

I have listed my index.ts, tsconfig.json and package.json below.
I will also release a better documentation of how to implement the SDK and example implementations in a few weeks time.

I will be in Holidays for the next two weeks starting next Monday, i will tell my co-workers to keep an eye out for this thread though if you have any questions/feedback.

```typescript
import express from 'express'
const app = express()
const port = 8080

import {
    DomainRobot,
    DomainRobotModels,
    DomainRobotResult,
    JsonResponseDataDomain
} from 'js-domainrobot-sdk'

app.get('/domainlist', async (req, res) => {
    let domainRobot = new DomainRobot({
        url: 'http://dev-proxy-lab.intern.autodns-lab.com:10025',
        auth: {
            user: 'root',
            password: 'test',
            context: 1,
        },
    })

    let queryFilter = new DomainRobotModels.QueryFilter({
        key: 'sld',
        value: 'example%',
        operator: 'LIKE',
    })
    let query = new DomainRobotModels.Query({
        filters: [queryFilter],
        view: {
            children: true,
        },
    })

    let result: DomainRobotResult<JsonResponseDataDomain, Number>

    try {
        result = await domainRobot.domain().cancelationList(query);
        //result = await domainRobot.domain().info("blub.de")
        console.log(result);
    } catch (DomainRobotException) {
        result = DomainRobotException;
    }

    res.send(result)
})

// start the Express server
app.listen(port, () => {
    console.log(`server started at http://localhost:${port}`)
})
```

my tsconfig.json
```json
{
    "compilerOptions": {
        "module": "commonjs",
        "esModuleInterop": true,
        "target": "es6",
        "noImplicitAny": true,
        "moduleResolution": "node",
        "sourceMap": true,
        "outDir": "dist",
        "baseUrl": ".",
        "allowJs": true,
        "paths": {
            "*": [
                "node_modules/*"
            ]
        }
    },
    "include": [
        "./src", "index.ts"
    ],
    "exclude": [
        "./node_modules",
        "./vendor",
        "./tests",
        "./tests_js"
    ]
}
```

and my package.json
```json
{
    "name": "ts-domainrobot-sdk",
    "main": "dist/index.js",
    "scripts": {
        "build": "tsc",
        "prestart": "yarn build",
        "watch": "nodemon --watch src --watch index.ts --exec yarn start",
        "start": "node ."
    },
    "version": "1.0.0",
    "license": "MIT",
    "dependencies": {
        "express": "^4.17.1",
        "js-domainrobot-sdk": "^1.2.2",
        "nodemon": "^2.0.4"
    },
    "devDependencies": {
        "@adelholtz/eslint-config": "https://github.com/adelholtz/eslint-config.git",
        "@types/express": "^4.17.7",
        "@types/node": "^14.0.23",
        "typescript": "^3.9.6"
    }
}
```

**Comment by patrickhilker** [comment]
Thank you very much for all the work you have done! Now there are only two types missing:

https://github.com/InterNetX/js-domainrobot-sdk/blob/1bd23f2/index.d.ts#L121
https://github.com/InterNetX/js-domainrobot-sdk/blob/1bd23f2/index.d.ts#L92

If you provide these types, it seems that nothing is missing. I will try to generate a client in the near future and implement it in our system. Then I will give more detailed feedback.

**Comment by adelholtz** [comment]
Hi

For this i cant provide a closer explanation of the type for data type. I use the ' extends Result ' approach to define the data for each special case because  the data structures heavily varies for each single API call.
There shouldn't be a scenario in which you have to use 'Result' directly and in my Typescript test project which i posted in a reply above i tried this and it worked as it was supposed to be.

```typescript
export interface Result {
    stid: String;
    status: ResponseStatus;
    object?: ResponseObject;
    ctid?: String;
    data: Array<object>;
}

// Service Response Definitions
export interface JsonResponseDataContact extends Result {
    data: DomainRobotModels.Contact[];
}
```

I updated the other type for messages in line 92 and published a new version 1.3.3
